<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Meta-binning : a pipeline to reconstruct genome sequence from metagenomes">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Meta-binning</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/mingleiR/meta-binning">View on GitHub</a>

          <h1 id="project_title">Meta-binning</h1>
          <h2 id="project_tagline">a pipeline to reconstruct genome sequence from metagenomes</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/mingleiR/meta-binning/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/mingleiR/meta-binning/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">

<hr>
<h1>A pipeline to binning metagenomic sequence based KDE and normal mixture model</h1>
<hr>
<p>A typical process of metagenomics analysis includes sampling, sequencing, data cleaning, assembly and gene calling. Because of the incomplete nature of sampling and internal bias in sequencing, it is difficult to recovery the full genomes of each genome from the metagenomics. Binning provides necessary complementation in practice.</p>
<p>In metagenomics, binning is a process of grouping raw reads or assembled contigs and assigning them to operational taxiomic unit. Here, we mean the contigs assembled from the raw reads. There are two basic methods of binning sequence, and one is based on composition features, the other is based on sequence similarity. We present an effective binning approach combined two strategies above. In addtion, an operable and detailed step-by-step guide is provided.</p>
<ul>
<li>Step 1. Collect important features value
<ul>
<li>1-1. Trimming</li>
<li>1-2. De novo Assembly</li>
<li>1-3. Extraction of selected fetures data</li>
</ul>
</li>
<li>Step 2. Filter the metagenomic assembly</li>
<li>Step 3. Binning the metagenomic assembly using selected features
<ul>
<li>3-1. Compute the KDE using the weighted data and create contour graph</li>
<li>3-2. Determine the initial group of data based on different level values of contour</li>
<li>3-3. Fit normal mixture model for the assembly</li>
<li>3-4. Validate each genome bin</li>
</ul>
</li>
<li>Step 4. Taxonomy assignment and functional analysis for each genome bin.
<ul>
<li>4-1. Toxonomy anslysis using MEGAN</li>
<li>4-2. Phylogenetic anlysis based on 16S rRNA sequence</li>
<li>4-3. Clusters of Orthologous Groups (COG) analysis</li>
<li>4-4. Pathway analysis using GhostKOALA service</li>
</ul>
</li>
</ul>
<h2><a id="Step_1_Collect_important_features_value_29"></a>Step 1. Collect important features value</h2>
<p>After getting the sequencing data, we can utilize a lot of open source tools to analyze the data, including trimming, assembling and annotation. The following step is for reference only.</p>
<h3><a id="11_Trimming_32"></a>1-1. Trimming</h3>
<p>Although there are many useful toolkits to perform the quality control, We choose Trimmomatic to clean our sequencing datasets. The detailed parameters is following:</p>
<pre><code>  java -jar trimmomatic-<span class="hljs-number">0.32</span><span class="hljs-class">.jar</span> -threads <span class="hljs-number">8</span> -phred33 -trimlog PE_trim<span class="hljs-class">.log</span> \
        PE_1<span class="hljs-class">.fastq</span> PE_2<span class="hljs-class">.fastq</span> \  
        PE_1_trimmed_paired<span class="hljs-class">.fq</span> PE_1_trimmed_unpaired<span class="hljs-class">.fq</span> \  
        PE_2_trimmed_paired<span class="hljs-class">.fq</span> PE_2_trimmed_unpaired<span class="hljs-class">.fq</span> \  
        ILLUMINACLIP:<span class="hljs-variable">$HOME</span>/Software/Trimmomatic-<span class="hljs-number">0.32</span>/adapters/Index/TruSeq2-PE-GCCAAT<span class="hljs-class">.fa</span>:<span class="hljs-number">2</span>:<span class="hljs-number">30</span>:<span class="hljs-number">10</span>:<span class="hljs-number">1</span>:true \  
        LEADING:<span class="hljs-number">3</span> TRAILING:<span class="hljs-number">3</span> MAXINFO:<span class="hljs-number">40</span>:<span class="hljs-number">0.4</span> MINLEN:<span class="hljs-number">20</span>  
</code></pre>
<p>The dataset generated from the mate-pair library was trimmed like above commands.</p>
<h3><a id="12_De_novo_Assembly_46"></a>1-2. De novo Assembly</h3>
<p>There are two types of sequence assembly, that is overlap/layout/consensus approach and de Bruijn graph approach. Because of the characteristics of next-generation sequencing platform and the incompletion of reference genome, de nove assembly is appropriate for metagenomics assembly. According to our experience, we use SPAdes Genomes Assembler to assembly our data. Executinng SPAdes as the following command:</p>
<pre><code>  spades<span class="hljs-class">.py</span> --dataset data<span class="hljs-class">.yaml</span> -t <span class="hljs-number">8</span> -o ./ -k <span class="hljs-number">21</span>,<span class="hljs-number">35</span>,<span class="hljs-number">51</span>,<span class="hljs-number">69</span> --carefull
</code></pre>
<p>Before using the commands above, you need to create your data.yaml file as the SPAdes online manual told.</p>
<h3><a id="13_Extraction_of_selected_fetures_data_55"></a>1-3. Extraction of selected fetures data</h3>
<p>In order to bin metagenomics sequence, we have to collect some fetures about the assembled fragments. For each contigs, GC content, length and coverage are extracted from assembly file using our R script. The script outputs the file <code>contigs.info</code>  including 4 column, that is contig name, contig length, GC content and coverage.</p>


      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Meta-binning maintained by <a href="https://github.com/mingleiR">mingleiR</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
